lang_qc/util/qc_state_helpers.py

```diff
@@ -0,0 +1,235 @@
+            )
+        )
+    ).scalar_one_or_none()
+
+
+def construct_seq_product_for_well(
+    run_name: str, well_label: str, qcdb_session: Session
+):
+    """Construct a SeqProduct for a well and push it to the database.
+
+    This assumes a 1-1 mapping between SeqProduct and SubProduct.
+
+    Args:
+        run_name: The run name.
+        well_label: The well label.
+        qcdb_session: A SQLAlchemy Session connected to the QC database.
+
+    Returns:
+        The SeqProduct which has been pushed to the QC database.
+    """
+
+    seq_platform = qcdb_session.execute(
+        select(SeqPlatform).where(SeqPlatform.name == "PacBio")
+    ).scalar_one_or_none()
+    if seq_platform is None:
+        raise Exception("PacBio SeqPlatform is not in the QC database.")
+
+    run_name_product_attr = qcdb_session.execute(
+        select(SubProductAttr).where(SubProductAttr.attr_name == "run_name")
+    ).scalar_one_or_none()
+    if run_name_product_attr is None:
+        raise Exception("PacBio run name SubProductAttr is not the QC database.")
+
+    well_label_product_attr = qcdb_session.execute(
+        select(SubProductAttr).where(SubProductAttr.attr_name == "well_label")
+    ).scalar_one_or_none()
+    if well_label_product_attr is None:
+        raise Exception("PacBio well label SubProductAttr is not in the QC database.")
+
+    seq_product = SeqProduct(
+        id_product=create_id_product(run_name, well_label),
+        seq_platform=seq_platform,
+        product_layout=[
+            ProductLayout(
+                sub_product=SubProduct(
+                    sub_product_attr=run_name_product_attr,
+                    sub_product_attr_=well_label_product_attr,
+                    value_attr_one=run_name,
+                    value_attr_two=well_label,
+                    properties=create_well_properties(run_name, well_label),
+                    properties_digest=create_well_properties_digest(
+                        run_name, well_label
+                    ),
+                )
+            )
+        ],
+    )
+
+    qcdb_session.add(seq_product)
+    qcdb_session.commit()
+
+    return seq_product
+
+
+def update_qc_state(
+    qc_status_post: QcStatus, qc_state_db: QcState, qcdb_session: Session
+):
+    """Update the properties of the QcState, without pushing the changes.
+
+    Args:
+        qc_status_post: The object containing the new properties to update.
+        qc_state_db: The object on which to apply the updates.
+        qcdb_session: A SQLAlchemy Session connected to the QC database.
+
+    Returns:
+        None
+    """
+
+    # Check that values are in the DB.
+    desired_qc_state_dict = qcdb_session.execute(
+        select(QcStateDict.id_qc_state_dict).where(
+            QcStateDict.state == qc_status_post.qc_state
+        )
+    ).one_or_none()
+    if desired_qc_state_dict is None:
+        raise Exception(
+            "Desired QC state is not in the QC database. It might not be allowed."
```

Comment:
As above, you could put the `qc_status_post.qc_state` into this error and the user would be certain of what they asked for. It sounds self-evident but it's really useful when the client is confused.